### DISCUSSION TITLE
Some particles are not decompressing correctly

### DISCUSSION THREAD

**Discussion by HassamChaudhary1** [axmolengine/axmol#1509] [Q&A] [answered]
ParticleSystem.cpp class is not able to decompressed below base64 STRING.
The base64 string is valid and I can decode it from python but when try to use this particle then ParticleSystem does not able to detect the right file type.

"H4sIAAAAAAAAA+VbB1RUR9ueu33pvUlZpAqCgtKlSRFEBFFBseKyAlKl2lsMRjR2iYololFRo0YjKrGixhp7i5oEY4gaTYJJsMf55/XODdeNyWdycljO+S/nYWdvfZ+3z1yIj0duCIXoYIyRBiHT8PM1BX0C8zYgh6a4u7QBOTQBWwLfNiCHJmBHEN0G5NAEPAj6twE5NIFOBOltQA5NoBtBcRuQQxMIJXinDcihCcQRLGwDcmgCCQTVbUCO1gLDG6cSbFfb19Yh+o90MJzgEO+7ug7eVifCVtQfzH+M/wVXTj4BHY8iuMD7/nfy/9UxGZVF0ErctQnaEyj+AW/uk89zNEGD2j71c/7uvqZUDmkr8eZgSOBP4PUWduL4COmniI7HETTx9nPHuHO5696kQ7C1J+Xfmrw5WBEk4z/35HwbCikfEW8sppiK2U3MOy7kXceogbu3NUEsZucDmuDNAeagZQQp9DvnrwL8ui2BnwSz/slhFuVuTo9JeHoR8a7n8/amz7LWMG8O4PdzCVQ8m3GcRTyukJe0COSYzRmLKHcXuk/GO1eMW/yA8xdnzNYGN/zf1Jr/AsA3gGAdwVgqq5hykPC46mK2RsAnrFssp9yDCPQo5Gp6EFM9diCYThCMW3JAW+gLOFtHEZwkmMfjLKdcDShfE8zmJzOCGsq9N/1uSM/To7qS0/v0wqyPdMGv14G2wJuTBfwQ5uN3CJbiFlsbUs4Q17YUCoJPKHeYx9rT/ZYERlQHgAjM9vxhmI0XLgdomjfiycLFJPjpIILHmO3VralN4RNqMcS2O2bn7vso9/GYXcdwpefY0mtCCOoIRuCWHMjnrgnbq9dujjf4J8Qp2DuL8tpBEE7gSPmB33ajdjxOz1mA2TgOoDoBHSgJPicYRu8nwy35j19DNO3vHHeIT/BL8HPw2ymU2wmCoZRbJGbnbwMIrtDjmzEbK7AfckYewUHM1n873BL/HH+u/mkq5rlnczWIq2FcXrPArJ8vpPyAJ6xTDCQYSVCA2X4WtsMEhQSZmO314NxNBIEENpj1fz2qVxn+c+/X2rz5/s71LWAbLpe3I3DCrJ9XU47fEWzArD9AX3Ofp5fZBGsJ7mLW5pAzfDBb00GHkC91eNw539eUr/N7Nn4tAztB3oYexI8ghuAzyvMXgi8w2ws8pPsaCeoJfiC4jtk+CWICcoInvZcpvT9X87m89zZznv+St3qvLqXcwS/B5lCnnKjckKv7YtbPz1GuLzDrA094+mimuthFMI2eD3Ud3l1AnrSgeuX8nst7rcmd7+sCKgOX2zl/hxiFXA0274HZGId4Bj+/jVu23+nnS4KnBBcx2++8T5BP0A+zPR/URSt6b77f8+tra3Ln4h780A6ztQt4DsbsvLwcs/3NVoIjmI1p6Hke8jjzN/CFx/Q45IEGgsuYrXN76X3WE1RhtibOxGwdgNwIeRNq6jDMroNBvYB10FDM1lOomx5UTugd9CkQ1SP4KcQn5NZ4zNYkqDWTMdujrsbs2tphKtP3BI/ewKG1N5j/g57OY3b9C2SE3Ar1BeoKrP9D3wg1FObaEEPgR9BjAm+wH+evtvQY6An0AD0H6DARszqFHgt0XIpZnYP+ITeBPSBPf0ywB7O5GuQBe4Ofgy0hrsG3X7yBw3OCnzGbBy5THhsJ5hOMoc+Fvj+Aygey8ut8a+Y7/twSch3kIKhDkOO5ePenuk6l+oc4hj6W85eXavwbqP4qCHIJkqj+Xeg94f7a+PU5rvq6Rmtw5/TNzde4fg7sYU3lhRzQHbM5C7hA/P9IeX5D8IynAy7nnyKYhNmetidma7wDZnMdV+O5uXFr53n1tUR+fYcaZ0TlhLoEOQZiBmoVzFfA98Hfz2A2d/1C+UIcQM3/mX6H+R3EZxi9B8xvwJ8gLrl5/ZvWdVqLP9/vOdtzfTzkEQVm/R5sDzY/i9kcDjkb8vMczOZLTHUAfR3kcehzoA4swWxegnuAHxnQ+8vo8zjureXrfPBrK9fPa1MZjai80N9AXwMxDjaHWIb6kY3Z+L9GuUMehPoP/Qz0NV9iNvdNoTqEvkYf/znWNTWfUe/tuPkrt1YBPgp9Cfg29CzQ10CtgJoB/gx9wFHKHfSSQo/DPA96IVj/gRyQSXUJca6FX7e5Jufv/DV3bn2OW6uBng7q3gGCCQSdCbpSfcBaTDTVC2xX6D7ofyFGuFiBOQDEyQDcYnP+endr836T76vbHvr4GZit9eDfMKeD+YgLPQa8IH8vo9zP0/3QU0Cc2GG214L3D9sIvsVsr8atYXNxrol45787Ul+zgt5oHOU1mPIwo1wUVAcOVA/vUu4H6T44DnnCnPKGugG+DjnwNGZ9gx/nmly/4J7LyQP9B+R06PV6U/m59WgTng6sKUZT7rsxm8/M6DVG9DpurQZ8HdZpIUb81fhqmjvwhpoOcQnvaKLw6+8idHHL3B54mVJdDKLcYc3KELfMMcDW3Bo95+fgUxMJFmM2d2h6rZbzfZAb8jfMGYKovfjvoDgdcO8l9CjHaMp9DW55b8Ffl+e/p+NiC/IHzFGcseZszgFkDsVs39qZyq2e/7g5vgy39L/Aswvlvgy32Jm/Hsl/N8n3syEEGZjNHZrkDfk5ktqB839+LuS/d+X6cA62lPsc3NKz8Dnz1+TUYxzew8Jc+5/+DcR/AQnVuzfl8KY5FV9+fl3mfEKXcp/J0xF3Lr+O8DnzZYB+Afpew1bkDbJDrlJgNjer9xn8HMi/Rr0mwifMZ8fx9qu/b/9fOQ1qJdTH1vo7fBFumU/+1Tl8+/Ptpf4+HeYzpW/g/U9qt2krcufmrm977l9xge+wRjMK/zme/6lMmq55b6MH9X2w/jfiDbrStKytAVhbTG0DcmgCsP6b2Abk0ARg/TuuDcihCUzG7JqepuXQBOD9R2AbkEMTgPU7jzYghyYAPfn/1/+RA3+3bQNyaALw/tOkDcihCcAalr4Gn69JcO+BNfJ8pMcgJEQIMSiU/GJ4Y8GrsQihkDpGSPeTkxkJHQvIL0M4DnuRjDHinWNKx+QoY8a7pzl3foiIseadn8S7Z98/nls/owhJEdKJIeOTCDYZ/WHoDznWKysvX2CAUG5ecWFSj+6KAQNTFdIz5E5yJEFeCKUpiwri+0b3g8tjoyIUReQk9Nr26MorBuiiR0yiQoH+2WaoLCgsJlInknGXdFWRkoynk3FOWXEB7G8iY5MR2TAWAHeTQiIgGVvAOIMdd3x1DjsOg3F6bl46GYPMBem56TA+QsbvlZaoyFjYi4zLS7NUZWR8iYwdc0pys8j4CVybq0orIurTgf3FKmUmGXcmY53CfkkRZNyNKFEngzcewRsXq8YUA6mI/IKxhVkZmcUKV2UHhVdAgL8iRlWWoyou9khMU2anFaYrIvJzC9LyxiLEcn61GYFuFUTJvl4Bvr4e3p5ePEX97cG33MC27OjXPq9sxpidbtn3pvPyqxHybya6mdeyb8RShHa9i5DF9ZZ9jh8ipE/sVnuWx8cM/CWzuLggsFOnsrIyzyyV0hMU+sf2P094i433PE+43R/qUUSqRqaV5BQrQG/K/Jz8kkJFUUGaUqXwUHfif33hm+XomKQaqSpU5ZErkomXZeVlEHPnpWcVZ+XnKbLy/sqI//IytY31a7IZr32JTIZ5IoOzJkj442kkMtZGwsEryRHmD7v1kicjiLwUuzus37/amD/fVTAXfhVlZby6LiKpn0JZUljKHoOwRGKkhfSRCbJEtsgBuSIP5I38UBAKQ1GoJ0pA/dBANBQpUSbKRYWoDE1AU1E5moXmocVoGVqF1qIatAVtR7vQXnQIfY5OoXPoMrqBGlAjuo+a0CP0nGEYKaPLGDOWjB3jxLgz3ow/E8JEMb2YJGYgM5zJYPKYEmYC8w4zi1nALGNWMzXMJ8we5hBzgjnPfMncYu4xvzDPBEKBjsBEYCNwFnQS+AvCBXGCfoIhggzBaME4wXTBHMESQbVgs6BWcEhwSnBZ0CC4L2gmCVxbaCa0F3oI/YURwgRhqnCksFA4SVghrBRWC7cI64THhBeFDcIHwqciichYpBB5iIJEMaL+IqVotGiSaLZomWiDqFZ0RHRRdEvUJHop1hVbi93FgeJY8QBxhrhMXC6uFK8T7xQfFV8WN4ofSSQSM4mLxE8SIxkoGSUZL5ktWSHZKjkoOS+5LWmWSqWWUndpsDRBmiYtlpZLl0o3Sw9IL0gbpU9k2jI7mbcsWpYqy5NNk1XKNsr2yy7I7sieyw3kTvJAeYI8XT5WPle+Vl4nPytvlD/XMtRy0QrW6qc1Smuq1hKtLVpHtW5q/aqtrd1eO0C7j3aW9hTtJdrbtI9r39J+qmOk46YToTNYp0Rnjs56nYM6X+r8qqur66wbppuqW6w7R7dG97Du17pP9Iz1PPVi9dL1JutV6dXqXdD7SV+u76Qfrj9Uf5x+pf4O/bP6DwzkBs4GEQZpBpMMqgz2GFw1aDY0NvQyTDDMNZxtuNHwhOFdI6mRs1GUUbrRdKM1RoeNbhsLjR2MI4yVxu8YrzU+atxoIjFxMYk1GWUyy+RjkzMmTaZGpl1Nk03HmFaZ7jNtMBOaOZvFmuWYzTXbbnbF7Jm5jXm4ucp8pvkW8wvmjy3aWYRZqCwqLLZaXLZ4ZqmwjLLMtpxvucvyKyuRlZtVH6syq5VWR60etDNpF9RO2a6i3fZ2160F1m7WSdbjrddYn7ZutrG16WFTYLPU5rDNA1sz2zDbUbaLbPfb3rMztguxy7JbZHfA7geFqSJckaNYojiiaLK3to+xL7FfbX/G/nl7l/b9209rv7X9Vw5aDv4OIx0WOdQ7NDnaOcY7TnDc5HjdSe7k75Tp9IHTMafHzi7OKc4znHc533WxcIl1GeeyyeWmq65rqOto12rXSx0kHfw7ZHdY0eGcm8DNxy3TrcrtrLvA3dc9y32F+/mO4o4BHfM6Vne86qHjEe5R6rHJ45anmWcvz2meuzx/6uTYKbXT/E7HOr3s7NM5p/Pazje8jLx6ek3zqvP6xdvNW+ld5X2pi26X6C6Tu+zu8rCre1dV15Vdr/kY+8T7zPCp9/nd18+30HeL7z0/R7/hfsv9rvqb+Cf6z/Y/HiAO6B4wOWBvwNNA38DiwO2BPwd5BGUHbQy6282lm6rb2m63g9sHpwWvDm4IUYQMD/kwpCHUPjQttDr02zCHsPSwdWF3wjuEjwrfHP5T987dC7vv7P44IjBiYsTBSGFkj8iKyDNRRlH9o5ZFfR3dPjojelN0Uw+fHuN7HIwRx8TFzI+5GmsTq4ytiW3q6ddzYs8jcTpxfeOWxX3by61XYa+6eEF8z/iF8Td7O/XO670rASXEJixM+CrRJXF04md9JH0S+1T1+T7JK2lC0rG+xn2H9d3Y91G/7v3m9rvR37V/Sf/6ZP3kwck1yY9TIlMWpDQM6DRg4oBTA60GZg3cnSpNTU5dl9o8KGrQ4kGNg30Glw++MsRlyJghJ4ZaDc0Zum+Y/rC0YTuGi4enDN84/EVaQlp1WvOI2BHLRzQpI5QfKO+nh6UvSr+nClYtUN0ZGTxywci7GcEZCzPuZYZmVmY+yIrIWpb1cFTMqFWjHmcnZK/PxjkpOVtzZbnDc/fkGeVl5x3Jt80fk3++wL2gvKBhdODoxaObCuMK1xUxRUOKdhebkGbqdIlrybslt0pDSqtKn5Qll+0YYzgmb8zpsW5jZ469My563EfjReOV4+sn2E+YOuHWxPCJqycxk0ZMqp/sMHn65MYpPaZsmKo1NXvqF9M6T1sw7bd3Ut6pm24zfcr02+/2eHdTuV55YfnVGUEzVr0nei/rvTMzu8xcOvNlRXrFyVmdZ1XOejFbOfvk+17vL3kfzxk558xc37kr50nm5c27Mj90/oYFhgvGLbi9MH5h7SLFoopFvy0etvhEZdfKVR9ofVDyQcOSXkt2L3VcOm/pi2WZyy5Xda/autx6+czlj1ekr7iwMmzlllU2q2atevZh1ofXVvdYXVvtXF25RrKmdM33a5PXHvvI/6OadVbrZq37fX3e+oYNSRuO1PjV1Gy03jh3k2BTyaZ7mwdvPvdx5Me7t3hsWb3VbOusbWhbybYfPhn+yZXtcdvrd/jv2PKp06fLdxrvrKhlasfWNu3K3NWwe+Du83t67qmvC6rb+ZnnZ+v32u+t2me6b+5+rf3T9+MD4w40Hyw4+OBQxqHb9cPqbxwecPjSkT5HzhyNO3r88+jPDx8LP3bgePDxvScCT+w56X9y1ynfU7WnfU7v/MLni51nfM/UnvU7u/tcwLm6893O778QeuHQxciLn1+KvXTqcu/L56/0v3Lt6uCrDdfSr939MufLh9dLrz+/MeWm+GbFVwZfVX5t/XX1Nx2+2drg27DvVuSt09/2/fbGbeXt+98Vffeicfr3ut9X3rG7U3PX++7ee9H3zv0w6IfG+wX3nz8o/9Hwx+U/uf706c9hP59uGtDU+LDwIf5l9q+Wv67/retv9c2JzV8/yn30/HHFE8snG576Pz32LOXZnedlL6Qvlvze4fe6l3Evb8LfV/wfwoniCBJLAAA="

```
bool isOK = image->initWithImageData(deflated.release_pointer(), imageDataLen, true);
AXASSERT(isOK, "CCParticleSystem: error init image with Data"); [Game Stopped here]
```

I debugged and figured out that, the small base64 string can perfectly decode but when passing long base64 string, it cause issuing.
Anyone can help me out in this?

**Comment by halx99** [comment, marked as answer]
it's `.tiff`, the `.tiff` support was removed since cocos2d-x-4.0

**Reply by HassamChaudhary1** [reply]
![myfile](https://github.com/axmolengine/axmol/assets/79073135/d00bfeaf-5621-4565-8cc2-4d1fba2fd1d8)
This is the output file and it's in .png format.

When I decode above (base64 compressed encoded) string then it's output file show the .png image.

**Reply by HassamChaudhary1** [reply]
`auto dataLen = textureData.size();`
Is it possible that the dataLen (6259+) is too big for this string? @halx99

**Reply by rh101** [reply]
@HassamChaudhary1 Stepping into `image->initWithImageData` with a debugger is the best way to figure out why it doesn't return the value that you expect.

**Reply by HassamChaudhary1** [reply]
`bool isOK = image->initWithImageData(deflated.release_pointer(), imageDataLen, true);`
Stepped into **initWithImageData**, figured out that the deflated data is incorrect. Seems like the decompressGZ is having issue with the long string.

deflated data is "MM", it should have something "/b/06PNG....." type string but it's not.

`auto deflated           = ZipUtils::decompressGZ(std::span{buffer});`

Debug output:
```
axmol: unsupported image format!
Assert failed: CCParticleSystem: error init image with Data
Assertion failed: (isOK), function initWithDictionary, file ParticleSystem.cpp, line 650.
```

@rh101

**Reply by halx99** [reply]
"MM" means it's really a `.tiff` image data, refer to: https://github.com/cocos2d/cocos2d-x/blob/v3/cocos/platform/CCImage.cpp#L675

**Reply by HassamChaudhary1** [reply]
To reproduce this, copy the base64 string (Above mentioned) and paste inside the class ParticleSystem.cpp line number 630.
Pass the hardcoded string like the attachment.

<img width="1512" alt="Screenshot 2023-12-12 at 5 32 02 PM" src="https://github.com/axmolengine/axmol/assets/79073135/d39c6c8a-fc2c-49a9-8da4-165782813e71">

**Reply by HassamChaudhary1** [reply]
I'm still not 100% sure that why it's decompressor output is in .tiff but when I try to decompressed on python, it's returns the .png image.
Do we have any other way to handle .tiff file?

**Reply by halx99** [reply]
Don't rely on python, it will decode tiff then reencode to .png

**Reply by HassamChaudhary1** [reply]
So, then how can I handle .tiff here? Any suggestions that I can do to make this string runfull? @halx99

**Reply by halx99** [reply]
if you really want support .tiff, just bring back from cocos2d-x-3.x, steps

1. add libtiff to axmol thirdparty
2. bring tiff init code from cocos2dx 3.x CCImage

**Reply by halx99** [reply]
[0001-tiff-support.patch](https://github.com/axmolengine/axmol/files/13651519/0001-tiff-support.patch)

1. `git apply 0001-tiff-support.patch`
2. `axmol build -p <target_platform> -xc '-DAX_WITH_TIFF=ON`

**Reply by HassamChaudhary1** [reply]
After applying the .patch file. TIFF file format code added. But when I try to compile I got below errors. @halx99 
<img width="979" alt="Screenshot 2023-12-13 at 11 14 39 AM" src="https://github.com/axmolengine/axmol/assets/79073135/ffd10e84-4c75-437b-8ffe-c56e0abeeddb">

**Reply by halx99** [reply]
I have tested on windows, not test on apple yet

**Reply by HassamChaudhary1** [reply]
Ok @halx99 . Let me know when you done test on apple.

**Reply by HassamChaudhary1** [reply]
I figured out the issue, it's causing bcoz "libtiff.a" library is missing. How can I include this lib in?
@halx99

**Reply by halx99** [reply]
it will auto link to app, I checked build target `osx`, works

**Reply by HassamChaudhary1** [reply]
It's not working. Please check your "AX_WITH_TIFF" value is defined? I think TIFF code (initWithTiffData) is not calling on your side.
It's compile fine when I set "AX_WITH_TIFF" value to 0.
Once this code is execute **initWithTiffData**, I face above lib issue. @halx99

**Reply by halx99** [reply]
the patch required AX_WITH_TIFF=1

**Reply by HassamChaudhary1** [reply]
I already added AX_WITH_TIFF with value 1. When I compile code with this, then it throws lib error for tiff. Error mentioned above already.
How can I add a new library into existing axmol test project?

**Reply by HassamChaudhary1** [reply]
2nd thing is, If I add manually file into cpp-test -> Build Phase -> "Link Binary with Libraries" section, then it works on simulator. But it's not a right way to add a lib in to project for only simulator. @halx99 
<img width="1076" alt="Screenshot 2023-12-13 at 4 37 12 PM" src="https://github.com/axmolengine/axmol/assets/79073135/2ca67ce6-c4f2-42be-828e-55e193084ead">

**Reply by halx99** [reply]
could you test target 'osx', the `ios` have cmake generate error from my test

**Reply by HassamChaudhary1** [reply]
Yes @halx99 , it's working fine on osx. How to make it fix on IOS side?

**Reply by halx99** [reply]
cmake issue, will fix patch soon

**Reply by HassamChaudhary1** [reply]
Ok thanks @halx99

**Reply by halx99** [reply]
After PR #1515 merged, update your repo and apply the patch: [0001-add-tiff-support.patch](https://github.com/axmolengine/axmol/files/13664402/0001-add-tiff-support.patch)

**Reply by HassamChaudhary1** [reply]
@halx99 It's working now. Thanks.

**Comment by halx99** [comment]
set to 0 means not enable tiff suport

获取Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
